wire-ios-data-model/Source/Core Crypto/CoreCryptoKeyProvider.swift

```diff
@@ -17,96 +17,252 @@
+            WireLogger.coreCrypto.info("Rotating core crypto key...", attributes: .safePublic)
+            try await rotateKey(path: path)
         } catch {
-            if createIfNeeded {
-                return try createCoreCryptoKey()
-            } else {
+            WireLogger.coreCrypto.warn(
+                "Failed to rotate core crypto key: \(String(describing: error))",
+                attributes: .safePublic
+            )
+            throw error
+        }
+    }
+
+    private func rotateKey(path: String) async throws {
+
+        // Get the old key, to rotate with the new key
+        guard let oldKey = try fetchCoreCryptoKey() else {
+            throw Error.keyNotFound
+        }
+        let oldKeyId = uniqueKeyId
+        let oldItem = CoreCryptoKeychainItem(uniqueKeyId: oldKeyId, userID: userID)
+        logRotation("fetched the old key")
+
+        // Generate a new key and save it with a new key ID
+        let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
+        let newKeyId = UUID()
+        let newItem = CoreCryptoKeychainItem(uniqueKeyId: newKeyId, userID: userID)
+        try KeychainManager.storeItem(newItem, value: newKey)
+        logRotation("generated and saved the new key")
+
+        do {
+            // Update the database
+            try await coreCryptoKeyMigrationManager.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+            logRotation("updated the database with the new key")
+
+            // Delete the old key
+            deleteOrMarkAsStale(item: oldItem)
+            logRotation("deleted or marked old key as stale")
+
+            // Update the unique key identifier
+            uniqueKeyId = newKeyId
+            logRotation("updated unique key identifier")
+
+            // Mark the rotation as done
+            coreCryptoKeyMigrationManager.markKeyRotationAsDone()
+            logRotation("marked key rotation as done")
+
+        } catch {
+
+            // Check if we need to clean up or rollback, otherwise throw the error
+            switch error {
+            case let CoreCryptoKeyMigrationManagerError.failedToUpdateKey(underlyingError):
+
+                // We failed to rotate the key, rollback by deleting the new key
+                deleteOrMarkAsStale(item: newItem)
+                logRotation("rollback: deleted or marked new key as stale")
+                throw underlyingError
+
+            default:
                 throw error
             }
         }
     }
 
-    public func updateDatabaseKey(path: String) async throws {
-        if let oldKey = try? fetchCoreCryptoKey() {
-            let item = CoreCryptoKeychainItem()
-            let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
-            try await coreCryptoKeyMigrationManager?.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+    private func logRotation(_ message: String) {
+        WireLogger.coreCrypto.info("[key rotation] \(message)")
```

Comment:
Question: does this need to be logged on prod?

```suggestion
        WireLogger.coreCrypto.info("[key rotation] \(message)", attribures: .safePublic)
```

Reply:
wasn't sure if it was worth flooding the logs, but I made it public for now and we can remove it later if needed [3dd71ad](https://github.com/wireapp/wire-ios/pull/3637/commits/3dd71ad89cadc9edf4040d625e88dd5aea8f9827)